rfcs/text/0003-web-rpc-hooks.md

```diff
@@ -0,0 +1,280 @@
+  return <div>Hello {user.name}</div>
+}
+```
+
+# Motivation
+
+Hooks and suspense provide opportunities for improved ergonomics of APIs for 
+data fetching and state management. We can take advantage of these best by
+updating our APIs to be compatible with hooks and suspense.
+
+# Detailed design
+
+This RFC proposes a few different APIs at different layers of abstractions. From lowest to highest level of abstraction:
+
+1. `import {useRPC} from 'fusion-plugin-rpc';
+2. `import {useRPCRedux} from 'fusion-plugin-rpc-redux-react';
+3. `import {createRPCResource} from 'fusion-plugin-rpc-redux-react'; (experimental)
+4. `import {createRPCReactorResource} from 'fusion-plugin-rpc-redux-react'; (experimental)
+
+
+## `useRPC`
+
+This API is not designed for direct consumer usage, but rather to be used as a 
+building block into other APIs. 
+
+```js
+import {useRPC} from 'fusion-plugin-rpc';
+
+function Component() {
+  const rpc = useRPC();
+  rpc.getUser(args); // returns a promise
+}
+```
+
+This can be implemented either via a js Proxy or by serializing the supported methods on the server into the html.
+
+## `useRPCRedux`
+
+This can be built using the `useRPC` hook and the `useDispatch` hook. This provides an abstraction that should be easy to use and 
+understand, but does require a small amount of boilerplate. It is designed to be used along side the useSelector API. 
+
+```js
+import {useRPCRedux} from 'fusion-plugin-rpc-redux-react';
+import {useDispatch, useSelector} from 'react-redux'; 
+
+function Component() {
+  const rpc = useRPCRedux();
+  const user = useSelector(state => state.user.user);
+  if (!user) {
+    throw rpc.getUser();
+  }
+  return <div>Hello {user.name}</div>
+}
+```
+
+It also supports doing multiple calls in parallel.
+
+```js
+import {useRPCRedux} from 'fusion-plugin-rpc-redux-react';
+import {useDispatch, useSelector} from 'react-redux'; 
+
+function Component() {
+  const rpc = useRPCRedux();
+  const user = useSelector(state => state.user.user);
+  if (!user) {
+    throw Promise.all(rpc.getUser(), rpc.getTrips());
+  }
+  return <div>Hello {user.name}</div>
+}
+```
+
+The drawback of this API is it requires users to separately check the cached state via useSelector. This is a few extra lines of code,
+however I think it is pretty straight forward.
+
+## `createRPCResource` (experimental)
+
+This API is experimental and should likely not be implemented until further knowledge of suspense data fetching patterns is available. 
+This pattern is inspired by the logic in the react-cache package, which shows some hints of how suspense based data fetching may look.
+
+A major benefit of this approach is it provides a unified approach for managing cached state, error states, and loading states. A drawback 
+to this approach is it takes more effort to get static typing.
+
+```js
+import {createRPCResource} from 'fusion-plugin-rpc-redux-react';
+
+const useUser = createRPCResource('getUser', state => state.user.user);
```

Comment:
The difference with this pattern and the react-cache library is we need access to hooks (useRPC, useDispatch, useSelector, etc) inside the resource. We could have this return an object with a read method like the react-cache library, but it would be hiding the fact that the read method would be using hooks. And technically, any function that uses hooks is a custom hook, so it should follow hook conventions